compiler/rustc_metadata/src/creader.rs

```diff
@@ -157,6 +157,43 @@ impl<'a> std::fmt::Debug for CrateDump<'a> {
     }
 }
 
+/// Reason that a crate is being sourced as a dependency.
+#[derive(Clone, Copy)]
+enum CrateOrigin<'a> {
+    /// This crate was a dependency of another crate.
+    Dependency {
+        dep_root: &'a CratePaths,
+        /// Dependency info about this crate.
+        dep: &'a CrateDep,
+    },
+    /// Injected by `rustc`.
+    Injected,
+    /// An extern that has been provided with the `force` option.
+    ForcedExtern,
+    /// Part of the extern prelude.
+    ExternPrelude,
+    /// Provided by `extern crate foo`.
+    AstExtern,
```

Comment:
Do we actually need to distinguish between `ForcedExtern`, `ExternPrelude` and `AstExtern`? If not I did prefer merging them into a single variant.

Reply:
They don’t have any behavior difference, so I’ll merge them

Reply:
Combined these into a single `Extern` variant.